lightning/src/chain/package.rs

```diff
@@ -0,0 +1,1037 @@
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	per_commitment_key,
+	weight,
+	amount,
+	htlc
+});
+
+/// A struct to describe a HTLC output on a counterparty commitment transaction.
+///
+/// HTLCOutputInCommitment (hash, timelock, directon) and pubkeys are used to generate a suitable
+/// witnessScript.
+///
+/// The preimage is used as part of the witness.
+#[derive(Clone, PartialEq)]
+pub(crate) struct CounterpartyOfferedHTLCOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	preimage: PaymentPreimage,
+	htlc: HTLCOutputInCommitment
+}
+
+impl CounterpartyOfferedHTLCOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, preimage: PaymentPreimage, htlc: HTLCOutputInCommitment) -> Self {
+		CounterpartyOfferedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			preimage,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyOfferedHTLCOutput, 33*3 + 32 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	preimage,
+	htlc
+});
+
+/// A struct to describe a HTLC output on a counterparty commitment transaction.
+///
+/// HTLCOutputInCommitment (hash, timelock, directon) and pubkeys are used to generate a suitable
+/// witnessScript.
+#[derive(Clone, PartialEq)]
+pub(crate) struct CounterpartyReceivedHTLCOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	htlc: HTLCOutputInCommitment
+}
+
+impl CounterpartyReceivedHTLCOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, htlc: HTLCOutputInCommitment) -> Self {
+		CounterpartyReceivedHTLCOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			htlc
+		}
+	}
+}
+
+impl_writeable!(CounterpartyReceivedHTLCOutput, 33*3 + HTLC_OUTPUT_IN_COMMITMENT_SIZE, {
+	per_commitment_point,
+	counterparty_delayed_payment_base_key,
+	counterparty_htlc_base_key,
+	htlc
+});
+
+/// A struct to describe a HTLC output on holder commitment transaction.
+///
+/// Either offered or received, the amount is always used as part of the bip143 sighash.
+/// Preimage is only included as part of the witness in former case.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderHTLCOutput {
+	preimage: Option<PaymentPreimage>,
+	amount: u64,
+}
+
+impl HolderHTLCOutput {
+	pub(crate) fn build(preimage: Option<PaymentPreimage>, amount: u64) -> Self {
+		HolderHTLCOutput {
+			preimage,
+			amount
+		}
+	}
+}
+
+impl_writeable_len_match!(HolderHTLCOutput, {
```

Comment:
We shouldn't use `len_match`, I dont think, for structs that are contained in larger structs contained in larger structs, if anything its a pessimism instead of an optimization.

Reply:
Ah good to know, updated for HolderHTLCOutput/HolderFundingOutput!